### PR TITLE
ENH: add n-dimensional DCT and IDCT to fftpack

### PR DESCRIPTION
**Pull request by grlee77**
I recently needed an n-dimensional `dct` and `idct`.  This PR adds these.  

The implementation is trivial, as the existing 1D routines already had `axis` support, but I thought it may still be worth providing.

Unlike for the n-dimensional FFT routines, there is no underlying n-d implementation in the `fftpack` Fortran code.  This does not seem to be a big problem in terms of performance.  As is, it is similar or faster than `fftn` for real-valued inputs and a bit slower for complex-valued ones:

potential updates:

- [x] add `dstn` and `idstn` as well?
- [x] raise an error if duplicate axes are provided?

results from some quick speed tests in IPython:

```python
for shape in [(16, 16), (256, 256, 256)]:
    for dtype in np.float32, np.complex64:
        r = np.ones(shape, dtype)
        print("\n---------------------------")
        print("dctn for dtype {}, shape {}".format(dtype, shape))
        %timeit scipy.fftpack.dctn(r)
        print("\nfftn for dtype {}, shape {}".format(dtype, shape))
        %timeit scipy.fftpack.fftn(r)
```

```
---------------------------
dctn for dtype <class 'numpy.float32'>, shape (16, 16)
The slowest run took 4.43 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 60.2 µs per loop

fftn for dtype <class 'numpy.float32'>, shape (16, 16)
The slowest run took 9.49 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 69.5 µs per loop

---------------------------
dctn for dtype <class 'numpy.complex64'>, shape (16, 16)
10000 loops, best of 3: 131 µs per loop

fftn for dtype <class 'numpy.complex64'>, shape (16, 16)
10000 loops, best of 3: 67.3 µs per loop

---------------------------
dctn for dtype <class 'numpy.float32'>, shape (256, 256, 256)
1 loop, best of 3: 841 ms per loop

fftn for dtype <class 'numpy.float32'>, shape (256, 256, 256)
1 loop, best of 3: 1.43 s per loop

---------------------------
dctn for dtype <class 'numpy.complex64'>, shape (256, 256, 256)
1 loop, best of 3: 2.13 s per loop

fftn for dtype <class 'numpy.complex64'>, shape (256, 256, 256)
1 loop, best of 3: 1.49 s per loop
```

### CODE REVIEW

**Comment by grlee77** [comment]
Reduced the precision slightly in the tests to fix the Travis test failures.

**Comment by rgommers** [comment]
@grlee77 thanks! This fell through the review cracks, but would be good to get it merged in the near future. Would you mind rebasing?

For your potential updates I'd say yes 2x.

File: scipy/fftpack/__init__.py [hunk 1]
```diff
@@ -19,6 +19,8 @@
    irfft - Inverse of rfft
    dct - Discrete cosine transform
    idct - Inverse discrete cosine transform
+   dctn - n-dimensional Discrete cosine transform
+   idctn - n-dimensional Inverse discrete cosine transform
```
Comment (rgommers) [inline scipy/fftpack/__init__.py]: missing ``dstn``, ``idstn`` here and in ``__all__.extend`` below

File: scipy/fftpack/realtransforms.py [hunk 2]
```diff
@@ -22,6 +22,248 @@
...
+    x = np.asanyarray(x)
+    shape, axes = _init_nd_shape_and_axes(x, shape, axes)
+    for n, ax in zip(shape, axes):
+        x = dct(x, type=type, n=n, axis=ax, norm=norm, overwrite_x=overwrite_x)
+    return x
+
+
+def idctn(x, type=2, shape=None, axes=None, norm=None, overwrite_x=False):
+    """
+    Return multidimensional Discrete Cosine Transform of x along the specified
```
Comment (rgommers) [inline scipy/fftpack/realtransforms.py]: maybe remove "of x" and make this fit on one line

Comment (rgommers) [inline scipy/fftpack/realtransforms.py]: applies to all added functions

File: scipy/fftpack/tests/test_real_transforms.py [hunk 3]
```diff
@@ -3,9 +3,11 @@
 from os.path import join, dirname
 
 import numpy as np
-from numpy.testing import assert_array_almost_equal, assert_equal
+from numpy.testing import (assert_array_almost_equal, assert_equal,
+                           assert_raises)
```
Comment (rgommers) [inline scipy/fftpack/tests/test_real_transforms.py]: This is just being fixed in gh-7777, so better ``from pytest import raises as assert_raises``

**Review by rgommers — CHANGES_REQUESTED** [review]
This all looks good to me, only a couple of minor comments.

**Comment by grlee77** [comment]
Thanks for the review @rgommers.  let me know if you want me to squash the commits

File: scipy/fftpack/realtransforms.py [hunk 4]
```diff
@@ -22,6 +22,244 @@
...
+    Notes
+    -----
+    For full details of the IDST types and normalization modes, as well as
+    references, see `idst`.
+
+    Examples
+    --------
+    >>> from scipy.fftpack import dstn, idstn
+    >>> y = np.random.randn(16, 16)
+    >>> np.allclose(y, idstn(dctn(y, norm='ortho'), norm='ortho'))
```
Comment (rgommers) [inline scipy/fftpack/realtransforms.py]: Here ``dctn`` should be ``dstn``. That will fix the doctest failure.

**Comment by rgommers** [comment]
you can squash if you prefer, but 12 commits is not excessive from my point of view

**Comment by grlee77** [comment]
okay, should be good to go now

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
In it goes, thanks @grlee77!

**Pull request merged**